rollup/rollupsyncservice/rollupsyncservice.go

```diff
@@ -0,0 +1,415 @@
+	}
+
+	// default to genesis batch meta.
+	parentBatchMeta := &rawdb.FinalizedBatchMeta{}
+	if batchIndex > 0 {
+		parentBatchMeta = rawdb.ReadFinalizedBatchMeta(s.db, batchIndex-1)
+	}
+
+	chunks, err := s.convertBlocksToChunks(blocks, chunkBlockRanges)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to convert blocks to chunks, batch index: %v, chunk ranges: %v, err: %w", batchIndex, chunkBlockRanges, err)
+	}
+	return parentBatchMeta, chunks, nil
+}
+
+func (s *RollupSyncService) getChunkRanges(batchIndex uint64, vLog *types.Log) ([]*rawdb.ChunkBlockRange, error) {
+	if batchIndex == 0 {
+		return []*rawdb.ChunkBlockRange{{StartBlockNumber: 0, EndBlockNumber: 0}}, nil
+	}
+
+	tx, _, err := s.client.client.TransactionByHash(context.Background(), vLog.TxHash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get transaction, err: %w", err)
+	}
+
+	return s.decodeChunkBlockRanges(tx.Data())
+}
+
+// decodeChunkBlockRanges decodes chunks in a batch based on the commit batch transaction's calldata.
+func (s *RollupSyncService) decodeChunkBlockRanges(txData []byte) ([]*rawdb.ChunkBlockRange, error) {
+	const methodIDLength = 4
+	if len(txData) < methodIDLength {
+		return nil, fmt.Errorf("transaction data is too short")
+	}
+
+	method, err := s.scrollChainABI.MethodById(txData[:4])
+	if err != nil {
+		return nil, fmt.Errorf("failed to get method by ID, ID: %v, err: %v", txData[:4], err)
+	}
+
+	decoded, err := method.Inputs.Unpack(txData[4:])
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack transaction data using ABI: %v", err)
+	}
+
+	const expectedLength = 4
+	if len(decoded) != expectedLength {
+		return nil, fmt.Errorf("invalid decoded length, expected: %d, got: %v", expectedLength, len(decoded))
+	}
+
+	version, ok := decoded[0].(uint8)
+	if !ok {
+		return nil, fmt.Errorf("failed to cast version to uint8")
+	}
+	if version != batchHeaderVersion {
+		return nil, fmt.Errorf("unexpected batch version, expected: %d, got: %v", batchHeaderVersion, version)
+	}
+
+	chunks, ok := decoded[2].([][]byte)
+	if !ok {
+		return nil, fmt.Errorf("failed to cast chunks to slice of byte slices")
+	}
+
+	return DecodeChunkBlockRanges(chunks)
+}
+
+// getBlocksInRange retrieves blocks from the blockchain within specified chunk ranges.
+func (s *RollupSyncService) getBlocksInRange(chunkBlockRanges []*rawdb.ChunkBlockRange) ([]*types.Block, error) {
+	var blocks []*types.Block
+
+	latestBlockNumber := s.bc.CurrentBlock().Number().Uint64()
+
+	var maxRequestedBlockNumber uint64
+	for _, chunkRange := range chunkBlockRanges {
+		if chunkRange.EndBlockNumber > maxRequestedBlockNumber {
+			maxRequestedBlockNumber = chunkRange.EndBlockNumber
+		}
+	}
+
+	if latestBlockNumber < maxRequestedBlockNumber {
+		time.Sleep(180 * time.Second)
+		return nil, fmt.Errorf("local node is not synced up to the required block height: %v", maxRequestedBlockNumber)
+	}
+
+	for _, chunkRange := range chunkBlockRanges {
```

Comment:
Can we use `start,end := chunkBlockRanges[0].StartBlockNumber, chunkBlockRanges[len(chunkBlockRanges)-1].EndBlockNumber`?

Reply:
Yes.